konveyor/pelorus
# feat: Add PagerDuty Failure Exporter

Signed-off-by: Mateus Oliveira <msouzaol@redhat.com>

## Describe the behavior changes introduced in this PR

Add PagerDuty Failure Exporter to Pelorus.

| Variable | Required | Default Value |
|----------|----------|---------------|
| PROVIDER | yes | `pagerduty` |
| LOG_LEVEL | no | `INFO` |
| SERVER (tracker_api) | NOT USED | - |
| API_USER | NOT USED | - |
| TOKEN | yes | get it at `https://your_subdomain.pagerduty.com/api_keys` (Integration -> API Access Keys) |
| APP_LABEL | NOT USED | `app.kubernetes.io/name` |
| PROJECTS | NOT USED | - |
| PELORUS_DEFAULT_KEYWORD | no | `default` |
| PAGERDUTY_URGENCY | no | - |
| PAGERDUTY_PRIORITY | no | - |

**TODO write tests and docs:**
- [x] create PagerDuty account and subdomains for Pelorus tests (and add maintainers there)
- [x] think about use cases: is there some similar to projects in PagerDuty? (accessing 2 or more subdomains, for example) ANSWER: NO
- [x] write/update documentation for failure exporter in docs folder (add configuring PagerDuty section)
- [x] add PagerDuty logo to docs index and to the list of supported providers

## Linked Issues

Ensure the requirements items in here were fulfilled https://github.com/orgs/konveyor/projects/51/views/2?pane=issue&itemId=20613492

Resolves #848 

## Testing Instructions

To test PagerDuty API, run
```
curl --request GET \
--url https://api.pagerduty.com/incidents \
--header 'Accept: application/vnd.pagerduty+json;version=2' \
--header 'Authorization: Token token=your_token' \
--header 'Content-Type: application/json' | jq
```
[More details about the response](https://developer.pagerduty.com/api-reference/9d0b4b12e36f9-list-incidents).

we care about the following data
```yaml
"created_at": "2023-02-08T16:57:09Z", # Used for creation timestamp
"status": "triggered", # triggered, acknowledged, resolved;  used to check if it is resolved
"last_status_change_at": "2023-02-08T17:29:24Z", # Used for resolution timestamp
"priority": null, # null, P1, P2, P3, P4, P5 (objects {}); used as filter
"urgency": "high", # high, low; used as filter
"service": {"summary": "string"} # Used as label, to match which app the issue is related to
```

To test the python code, add this code to the end of `exporters/failure/collector_pagerduty.py` file
```python
if __name__ == "__main__":
    logging.getLogger().setLevel(logging.DEBUG)
    pager_duty = PagerdutyFailureCollector(token="your_token")
    pager_duty.search_issues()
```
and run
```
python exporters/failure/collector_pagerduty.py
```
